### ISSUE TITLE
MANNs with sparse reads and writes

### ISSUE THREAD

**Issue by rgobbel** [LeavesBreathe/tensorflow_with_latest_papers#13]
Obviously, we would all love to see an implementation of https://arxiv.org/pdf/1610.09027v1.pdf, the paper that describes amazing performance gains for training MANNs by using sparse associative memory.

**Comment by LeavesBreathe** [comment]
Hey @rgobbel Thanks for all the comments you've made recently. Right now I'm working on ByteNet  here: 

https://github.com/LeavesBreathe/bytenet_tensorflow

So I'm occupied doing this right now. Later on, sparse associative memory is an interest I have as well. Hopefully this can be done at a future date. Maybe the two can be integrated together :)

**Comment by rgobbel** [comment]
I'd be happy to collaborate on a SAM implementation, since it's something I
could make immediate use of. I'd be happy to contribute to the ByteNet
implementation as well, if there's some manageable chunk that makes
sense--let me know.

-Randy

**Comment by LeavesBreathe** [comment]
@rgobbel if you're interested in contributing to bytenet, I have some tasks to be done under the 'need to implement section'. Any of those tasks would be appreciated. A seq2seq training framework should be simple to pull from tensorflow.

**Comment by rgobbel** [comment]
A minimal Multiplicative Unit is really simple:

def mu(h, filter_width, padding, name='mul_unit'):
  initializer = tf.contrib.layers.xavier_initializer_conv2d()
  with tf.name_scope(name):
    channels = h.get_shape()[-1]
    filter_shape = [filter_width, filter_width, channels, channels]
    w1 = tf.get_variable('w1', initializer=initializer, shape=filter_shape)
    w2 = tf.get_variable('w2', initializer=initializer, shape=filter_shape)
    w3 = tf.get_variable('w3', initializer=initializer, shape=filter_shape)
    w4 = tf.get_variable('w4', initializer=initializer, shape=filter_shape)
    g1 = sigmoid(convolution(h, w1, padding), name='g1')
    g2 = sigmoid(convolution(h, w2, padding), name='g2')
    g3 = sigmoid(convolution(h, w3, padding), name='g3')
    u = tanh(convolution(h, w4, padding), name='u')
    return g1 * tanh(g2 * h + g3 * u)


How fancy does it need to be?

-Randy

**Comment by rgobbel** [comment]
Minimal Residual Multiplicative Block:

def rmb(h, name='residual_mult_block'):
  initializer = tf.contrib.layers.xavier_initializer_conv2d()
  with tf.name_scope(name, 'rmb', [h]) as scope:
    batch_size, _, _, channels = h.get_shape()
    w1 =  get_variable('w1', initializer=initializer, shape=[1, 1,
channels, channels/2])
    w4 = get_variable('w4', initializer=initializer, shape=[1, 1,
channels/2, channels])
    h1 = convolution(h, w1, 'SAME', name='h1')
    with tf.variable_scope('mult1'):
      h2 = mu(h1, 3, 'SAME', name='h2')
    with tf.variable_scope('mult2'):
      h3 = mu(h2, 3, 'SAME', name='h3')
    h4 = convolution(h3, w4, 'SAME', name='h4')
    return add(h, h4, name=scope)

**Comment by LeavesBreathe** [comment]
I will actually be integrating these in shortly, thanks for the functions. I will add the causal dilated convolutions. You're right - they are not hard to integrate. However, what will be more difficult will be getting the decoder portion functioning correctly. I"m finishing up the source network right now. I'll put these mult ints blocks in there.

**Comment by rgobbel** [comment]
The res-mult block was taken directly from the Pixel Net paper, so do take
a close look when integrating it. It ByteNet, it looks like the sub-batch
normalization actually happens inside of the decoder version of the
multiplicative unit. I also didn't make any attempt to implement the
masked--again not rocket science, but more than zero effort to get it right.

-Randy
